Config/FilterFactoryCollection.php

```diff
@@ -0,0 +1,57 @@
+<?php
+
+/*
+ * This file is part of the `liip/LiipImagineBundle` project.
+ *
+ * (c) https://github.com/liip/LiipImagineBundle/graphs/contributors
+ *
+ * For the full copyright and license information, please view the LICENSE.md
+ * file that was distributed with this source code.
+ */
+
+namespace Liip\ImagineBundle\Config;
+
+use Liip\ImagineBundle\Exception\Config\Filter\NotFoundException;
+use Liip\ImagineBundle\Factory\Config\FilterFactoryInterface;
+
+class FilterFactoryCollection
+{
+    /**
+     * @var FilterFactoryInterface[]
+     */
+    private $filterFactories = [];
+
+    /**
+     * @param FilterFactoryInterface ...$filterFactories
+     */
+    public function __construct(FilterFactoryInterface ...$filterFactories)
```

Comment:
got your point, but how usage of array and additional checks of type is better ?
since this object actually plays a role of collection and it is constructed with FilterFactoryInterface(s) why just not declare it directly ?

Reply:
as you loop over the parameter anyways, you could also do an instanceof check, and have the `@param FilterFactoryInterface[] $filterFactories` to allow code sniffers to detect errors. 

if we ever need to add an (optional) constructor argument, the current form will be a problem as we can't add anything else without a BC break.

Reply:
why would we ever need to to add something to constructor of this class ?
the purpose of this object is to store FilterFactoryInterfaces only and thats it. it shouldn't contain any logic. tbh, in this case specifically, I don't see advantages of array implementation.

Reply:
all factories are declared as `final`, does it sill make sense to mark them additionally with `@internal` ?

Reply:
yes, because `final` means you can't extend the class, but `@internal` means you should not interact with it at all, e.g. instantiate it in your code and so something with it.

Reply:
I guess within Symfony the pattern used most often is an `add()` method to build the collection instead of a constructor argument. obviously this has more performance overhead however. I think this is an interesting approach but since variadic in user land is still a new concept there is less experience with it.

but I wonder if it wouldn't be better to use a factory method instead of the constructor. ie. something like: 

```
class FilterFactoryCollection
{
    static public function createFromList(FilterFactoryInterface ...$filterFactories)
    {
        $collection = new self;
        $collection->addList(FilterFactoryInterface ...$filterFactories);
    }
}
```

this keeps the options open for if we ever do need to add anything to the constructor. while keeping the convenience and more or less the overhead reduction benefits.

Reply:
this object has only one purpose, to store the FilterFactoryInterfaces (which are clearly its dependencies and should be specified), I dont see any valid case which would lead to altering constructor. Also, I am not sure is it a good practice to do indirect object injections (via setters).

Reply:
ok lets keep it as proposed. like I said, a good chance of this feeling "weird" is simply the lack of experience with variadic in user land .. and honestly this is not the most "risky" place .. if we do in the end realize we need support additional constructor parameters .. we can then still use a static factory method for that case.